[MultiSelect] Enable pasting of multiple values (same as TagInput)

cmslewis opened the pull request:
#### (No issue)

### Checklist

- [x] Includes tests
- [x] Update documentation

<!-- DO NOT enable CircleCI for your fork. Our build will run when you open this PR. -->

### Changes proposed in this pull request:

Allow pasting of multiple values in `MultiSelect`.

#### Given:

The clipboard contains separated values (where the separator is configurable via `tagInputProps.separator`):

![image](https://user-images.githubusercontent.com/443450/54570304-4d785100-499b-11e9-9e64-55c31e367cc0.png)

#### You can now:

* Select multiple _existing_ items on paste:
    ![2019-03-18 16 29 21](https://user-images.githubusercontent.com/443450/54570337-6da81000-499b-11e9-9023-b83c9323409b.gif)
* _(If the "Create item" option is enabled)_
    Select a combination of _existing_ and _newly created_ items on paste:
    ![2019-03-18 16 34 22](https://user-images.githubusercontent.com/443450/54570440-b52e9c00-499b-11e9-9827-9a862cc3f4e2.gif)

* _(If the "Create item" option is enabled)_
    Select all _newly created_ items on paste:
    ![2019-03-18 16 36 55](https://user-images.githubusercontent.com/443450/54570524-10608e80-499c-11e9-8226-e85d0ace7238.gif)

* _(UX nicety)_
   If a pasted value is already selected, set that item as the active item on paste:
    _In general, the code moves the active item to the final existing item amongst the pasted values._
    ![2019-03-18 16 39 15](https://user-images.githubusercontent.com/443450/54570639-836a0500-499c-11e9-944f-70526381f835.gif)


### Reviewers should focus on:

I originally tried to implement this outside of `MultiSelect` in a wrapping component (by listening to `tagInputProps.onAdd`, which is invoked when values are added via paste), but that didn't work.

First, note that adding values via pasting does _not_ invoke our normal on-select listeners, but it does still invoke `tagInputProps.onAdd`. Thus, we have to explicitly invoke `onItemSelect` or similar from within `tagInputProps.onAdd` in order for pasted values to be emitted.

Unfortunately, when you emit `onItemSelect` from `tagInputProps.onAdd`, you end up double-invoking it in the default selection case, which leads to a bug. In particular, pressing <kbd>Enter</kbd> or clicking to select an item will emit `onItemSelect` from `tagInputProps.onAdd`, then by the time the `MultiSelect` change handler is invoked, the pasted value will already be selected, and it would  be toggled to _deselected_ immediately.

In this PR, I now have `TagInput`'s `onAdd` emit a second parameter of type `TagInputAddMethod`, which tells callers _how_ the emitted values were added (`default`, on `blur`, or on `paste`).¹  Only if the values were added via `paste` do I take special action to select matching items or create new ones if possible; otherwise, I don't do anything, and the `MultiSelect` component behaves just as before. The second parameter here can be safely ignored, which avoids API breaks.

---

¹ This consolidates logic, because parent components won't need to do any key listening to figure out if a particular keydown will trigger a paste.

❌ In that world, you might have to do something 🤡clowny like this:
```tsx
class MyWrapper extends React.PureComponent {
    private wasLastKeyDownAPasteOperation = false; 

    // render a MultiSelect, and pass tagInputProps: { onAdd: this.handleTagInputAdd } 
    private handleKeyDown(e) {
        this.wasLastKeyDownAPasteOperation = e.which === Keys.LETTER_V && e.metaKey;
    }
    private handleTagInputAdd(values: string) {
        if (this.wasLastKeyDownAPasteOperation) {
            // Then we know the values were added on paste, which would have
            // bypassed our normal key and click selection listeners. We'll need
            // to emit the values manually. 
        }
    }
}
```


### Screenshot

See above.

giladgray commented on packages/core/src/components/tag-input/tagInput.tsx:
```diff
@@ -15,6 +15,26 @@ import * as Utils from "../../common/utils";
+/**
+ * An enumeration of methods in which a `TagInput` value might have been added.
+ */
+export enum TagInputAddMethod {
```
I would prefer the `const/type` approach so users can use string literals instead of importing this long-named type in order to `switch` on it.

cmslewis replied:
@giladgray Product teams have requested string-literal enums in the past, but happy to do a simpler type. Is this what you mean?

```tsx
type TagInputAddMethod = "default" | "blur" | "paste";
```

Or perhaps just inline it?

```tsx
onAdd?: (values: string[], method: "default" | "blur" | "paste") => boolean | void;
```

adidahiya replied:
I like the first one,

```
export type TagInputAddMethod = "default" | "blur" | "paste";
```

cmslewis replied:
Done

giladgray commented on packages/select/src/common/listItemsProps.ts:
```diff
@@ -34,6 +34,13 @@ export interface IListItemsProps<T> extends IProps {
+     * to match pasted values to existing items, so that items can be emitted
+     * via `onItemsPaste`.
+     */
+    itemEqualsQuery?: (item: T, query: string) => boolean;
```
isn't this just `itemPredicate`? https://github.com/palantir/blueprint/blob/develop/packages/select/src/common/predicate.ts#L23

i don't see the need for a new prop here, except that it slightly changes the semantics of `itemPredicate` in that you may want to define `itemListPredicate` for fast querying _and_ `itemPredicate` for pasting, instead of one-or-the-other.

cmslewis replied:
@giladgray They're unfortunately different:
- `itemPredicate` is for filtering: an item matches if the query matches any substring of the item title (e.g.)
- `itemEqualsQuery` is for exact matching: an item matches iff the query === the full item title (e.g.) 

Possible to add an `exactMatch` flag or something in `ItemPredicate<T>`?

adidahiya commented on packages/core/src/components/tag-input/tagInput.tsx:
```diff
@@ -15,6 +15,17 @@ import * as Utils from "../../common/utils";
 
+/**
+ * A type reflecting the manner in which a `TagInput` value was added.
+ * - `"default"` - indicates that a value was added in the default fasion, via
```
nit: "fashion"

actually I would just simplify, remove the metaphor, and say "... was added by manual selection"

adidahiya commented on packages/core/src/components/tag-input/tagInput.tsx:
```diff
@@ -15,6 +15,17 @@ import * as Utils from "../../common/utils";
 import { ITagProps, Tag } from "../tag/tag";
 
+/**
+ * A type reflecting the manner in which a `TagInput` value was added.
```
nit: instead of "reflecting the manner ...", simpler language would be "The method in which a TagInput value was added."

adidahiya commented on packages/select/src/components/query-list/queryList.tsx:
```diff
@@ -54,6 +54,24 @@ export interface IQueryListRendererProps<T>  // Omit `createNewItem`, because it
+     * `createNewItemFromQuery`.
+     *
+     * If `itemEqualsQuery` returns multiple matching items for a particular
+     * `value`, then only the first matching item will be emitted.
```
do you mean "for a particular query"? if not, what is `value` here?

adidahiya commented on packages/select/src/components/query-list/queryList.tsx:
```diff
@@ -350,6 +369,40 @@ export class QueryList<T> extends React.Component<IQueryListProps<T>, IQueryList
+            this.setActiveItem(nextActiveItem);
+        }
+
+        // No need to update the active item.
```
what's this comment about? you _may_ have updated the active item at this point, no?

cmslewis replied:
Old cruft. Removed. 👍

adidahiya commented on packages/select/src/components/query-list/queryList.tsx:
```diff
@@ -350,6 +369,40 @@ export class QueryList<T> extends React.Component<IQueryListProps<T>, IQueryList
+        // Find an exising item that exactly matches each pasted value, or
+        // create a new item if possible. Ignore unmatched values if creating
+        // items is disabled.
+        const pastedItemsToEmit = values.reduce<T[]>((agg, value) => {
```
nit: I would slightly prefer this as a regular for loop, since that would be easier to read than a `reduce` which produces a list

adidahiya commented on packages/select/src/components/query-list/queryList.tsx:
```diff
@@ -443,6 +498,19 @@ function pxToNumber(value: string | null) {
 
+function getMatchingItem<T>(query: string, { items, itemEqualsQuery }: IQueryListProps<T>): T | undefined {
+    if (Utils.isFunction(itemEqualsQuery)) {
+        // .find() doesn't exist in ES5. Alternative: use a for loop instead of
```
AFAIK, `Array.prototype.find` should be supported in blueprint's environment.. can you elaborate here why that doesn't work?

cmslewis replied:
@adidahiya Yep, the compile step was failing on Circle. [Here is the failing build](https://circleci.com/gh/palantir/blueprint/24768?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) from a few commits back.

adidahiya replied:
that's unfortunate, I wish we had added it to the environment requirements. oh well.

adidahiya commented:
An `exactMatch` flag for `itemPredicate` does sound like it would work a little better, have you tried that? I think you should

cmslewis commented:
@adidahiya Thanks for the CR; I addressed your comments. The most recent commit includes a crack at using `itemPredicate` with an `exactMatch` flag. Two minor downsides are that the description for `itemPredicate` is now fairly complex, and attempting to handling pasting but not filtering will cause `QueryList` to invoke `itemPredicate` while typing a query when it otherwise wouldn't have. But that's probably a really uncommon case.

adidahiya commented on packages/select/src/common/predicate.ts:
```diff
@@ -14,10 +14,20 @@
+ * Customize querying of individual items.
  *
- * If defined with `itemListPredicate`, this prop will be ignored.
+ * __Filtering a list of items.__ This function is invoked to filter the list of
```
it's weird to have the exact same documentation block in two places. it's especially weird to reference the prop names here, when we're not in a react component class module... can you simplify this to a simple one-sentence description and tell people to "see its usage sites in `IListItemsProps` ..." ?

cmslewis commented:
@adidahiya Done! Now concatenating all unrecognized values into the input query on paste (only if the "Create item" option isn't enabled):
![2019-03-22 10 47 16](https://user-images.githubusercontent.com/443450/54843439-92e59880-4c91-11e9-99a8-9f837a35ea91.gif)
